trie-db/src/triedbmut.rs

```diff
@@ -711,18 +742,144 @@ impl<'db, L: TrieLayout> TrieDBMutBuilder<'db, L> {
+	// Storing the key and removed nodes (only needed for old trie).
+	pub key_childset: Option<(Vec<u8>, Vec<(H, OwnedPrefix)>)>,
+}
+
+#[derive(Debug)]
+pub struct ExistingChangesetNode<H, DL> {
+	pub hash: H,
+	pub prefix: OwnedPrefix,
+	pub location: DL,
+}
+
+#[derive(Debug)]
+pub enum ChangesetNodeRef<H, DL> {
+	New(NewChangesetNode<H, DL>),
+	Existing(ExistingChangesetNode<H, DL>),
+}
+
+impl<H, DL> ChangesetNodeRef<H, DL> {
+	pub fn hash(&self) -> &H {
+		match self {
+			ChangesetNodeRef::New(node) => &node.hash,
+			ChangesetNodeRef::Existing(node) => &node.hash,
+		}
+	}
+}
+
+#[derive(Debug)]
+pub struct Changeset<H, DL> {
+	pub root: ChangesetNodeRef<H, DL>,
+	pub removed: Vec<(H, OwnedPrefix)>,
+}
+
+pub fn prefix_prefix(ks: &[u8], prefix: Prefix) -> (Vec<u8>, Option<u8>) {
+	let mut result = Vec::with_capacity(ks.len() + prefix.0.len());
+	result.extend_from_slice(ks);
+	result.extend_from_slice(prefix.0);
+	(result, prefix.1)
+}
+
+impl<H: Copy, DL: Default> Changeset<H, DL> {
+	pub fn apply_to<K, MH>(&self, mem_db: &mut MemoryDB<MH, K, DBValue>) -> H
+	where
+		K: memory_db::KeyFunction<MH> + Send + Sync,
+		MH: Hasher<Out = H> + Send + Sync,
+	{
+		for (hash, prefix) in &self.removed {
+			mem_db.remove(hash, (prefix.0.as_slice(), prefix.1));
+		}
+
+		fn apply_node<H, DL, MH, K>(
+			node: &ChangesetNodeRef<H, DL>,
+			mem_db: &mut MemoryDB<MH, K, DBValue>,
+			ks: Option<&[u8]>,
+		) where
+			K: memory_db::KeyFunction<MH> + Send + Sync,
+			MH: Hasher<Out = H> + Send + Sync,
+		{
+			match node {
+				ChangesetNodeRef::New(node) => {
+					let ks = if ks.is_some() {
+						ks
+					} else if let Some((k, removed)) = node.key_childset.as_ref() {
+						for (hash, p) in removed.iter() {
+							let prefixed = prefix_prefix(k.as_slice(), (p.0.as_slice(), p.1));
+							mem_db.remove(hash, (prefixed.0.as_slice(), prefixed.1));
+						}
+						Some(k.as_slice())
+					} else {
+						None
+					};
+					for child in &node.children {
+						apply_node(child, mem_db, ks);
+					}
+					if let Some(ks) = ks {
+						let prefixed = prefix_prefix(ks, (node.prefix.0.as_slice(), node.prefix.1));
+						mem_db.insert((prefixed.0.as_slice(), prefixed.1), &node.data);
+					} else {
+						mem_db.insert((node.prefix.0.as_slice(), node.prefix.1), &node.data);
+					}
+				},
+				ChangesetNodeRef::Existing(_) => {},
+			}
+		}
+		apply_node::<H, DL, MH, K>(&self.root, mem_db, None);
+		self.root_hash()
+	}
+
+	pub fn root_hash(&self) -> H {
+		match &self.root {
+			ChangesetNodeRef::New(node) => node.hash,
+			ChangesetNodeRef::Existing(node) => node.hash,
+		}
+	}
+
+	pub fn empty(root: H) -> Self {
+		Self {
+			root: ChangesetNodeRef::Existing(ExistingChangesetNode {
+				hash: root,
+				prefix: (BackingByteVec::new(), None),
+				location: Default::default(),
+			}),
+			removed: Default::default(),
+		}
+	}
+
+	pub fn to_insert_in_other_trie(mut self, root_at: Vec<u8>) -> Box<ChangesetNodeRef<H, DL>> {
```

Comment:
changes from arkpar/new-trie. Probably not the best naming.

Reply:
True, might need to pass this `root_at` as param on commit (as Option<Vec<u8>>), or a different commit function when needed.
(root_at is only needed for rocksdb HashDB prefixing).